### PR TITLE
Secrets Manager and secrets specification for tasks

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
[wip]Pending adding secrets to flyteidl tasktemplate
This enables every python task or a python plugin to request for secrets and the secrets are retrieved in a standardized way during the execution using the flytekit context's new SecretsManager.

SecretsManager will resolve the secrets by first looking at Env Var and then at the configured secrets directory.
Rules:
- the env var should be uppercased and should be prefixed with the configured prefix. Default will be _FSEC_{secret_key}
- If env var is not found, then Secretsmanager will look for a file under {default_dir}/{prefix}{secret_key}
default_dir is set to /etc/secrets (can be overriden using an env var)
prefix is set to ""

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/796

### CODE REVIEW

**Review by wild-endeavor — DISMISSED** [review]
I feel like the user interface for secrets will evolve as people start using it (and maybe possibly as we add an authz layer).  Not sure how exactly though.

File: tests/flytekit/unit/core/test_type_hints.py [hunk 1]
```diff
@@ -1006,3 +1007,15 @@ def squared(value: int) -> typing.List[typing.Dict[str, int]]:
     def compute_square_wf(input_integer: int) -> typing.List[typing.Dict[str, int]]:
         compute_square_result = squared(value=input_integer)
         return compute_square_result
+
+
+def test_secrets():
+    @task(secret_requests=["my_secret"])
```
Comment (wild-endeavor) [inline tests/flytekit/unit/core/test_type_hints.py:1013]: Should this be `Secret("my_secret")`

Comment (kumare3) [reply tests/flytekit/unit/core/test_type_hints.py:1013]: aah yes it should be, and thus in this case maybe i should do a check on this

File: flytekit/core/base_task.py [hunk 2]
```diff
@@ -304,16 +311,21 @@ def __init__(
         name: str,
         task_config: T,
         interface: Optional[Interface] = None,
-        environment=None,
-        task_type_version=0,
+        environment: Optional[Dict[str, str]] = None,
```
Comment (cosmicBboy) [inline flytekit/core/base_task.py:314]: why are we removing `task_type_version`?

File: flytekit/common/tasks/sdk_runnable.py [hunk 3]
```diff
@@ -25,12 +25,53 @@
...
+            raise ValueError("Bad secrets key. Cannot be an empty string or None.")
+
+        env_var = self.get_secrets_env_var(secrets_key)
+        fpath = self.get_secrets_file(secrets_key)
+        v = os.environ.get(env_var)
+        if v is not None:
+            return v
+        if os.path.exists(fpath):
+            with open(fpath, "r") as f:
+                return f.read()
```
Comment (cosmicBboy) [inline flytekit/common/tasks/sdk_runnable.py:65]: maybe a `strip()` here in case of newlines??

File: flytekit/models/security.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+from flytekit.models import common as _common
+
+
+@dataclass
+class Secret(_common.FlyteIdlEntity):
+    class MountType(Enum):
+        ENV_VAR = _sec.Secret.MountType.ENV_VAR
+        """
+        Use this if the secret can be injected as an environment variable. Usually works for symmetric keys, passwords etc
```
Comment (cosmicBboy) [inline flytekit/models/security.py:15]: just a question: is this way of using multi-line strings as in-line comments a convention, or important for sphinx docs?

File: flytekit/core/base_task.py (hunk 2 above)
Comment (kumare3) [reply flytekit/core/base_task.py:314]: not removing, i am keeping it so that, if something is declared in the base, it does not need to repeated at every class layer - **kwargs should take care of it?

File: flytekit/common/tasks/sdk_runnable.py (hunk 3 above)
Comment (kumare3) [reply flytekit/common/tasks/sdk_runnable.py:65]: will do

File: flytekit/models/security.py (hunk 4 above)
Comment (kumare3) [reply flytekit/models/security.py:15]: Thats what I understand, that sphinx will convert this to a doc-string on the constant

File: flytekit/common/tasks/sdk_runnable.py (hunk 3 above)
Comment (kumare3) [reply flytekit/common/tasks/sdk_runnable.py:63]: done

File: tests/flytekit/unit/core/test_type_hints.py (hunk 1 above)
Comment (kumare3) [reply tests/flytekit/unit/core/test_type_hints.py]: done

File: flytekit/models/security.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        FILE = _sec.Secret.MountType.FILE
+        """
+        Use this for Secrets that cannot be injected into env-var or need to be available as a file
+        Caution: May not be supported in all environments
+        """
+
+    group: str
+    key: str
+    group_version: Optional[str] = None
+    mount_requirement: MountType = MountType.ENV_VAR
```
Comment (EngHabu) [inline flytekit/models/security.py:34]: Should this be ANY?

**Review by EngHabu — APPROVED** [review]

File: flytekit/configuration/secrets.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from flytekit.configuration import common as _common_config
+
+# Secrets management
+SECRETS_ENV_PREFIX = _common_config.FlyteStringConfigurationEntry("secrets", "env_prefix", default="_FSEC_")
+"""
+This is the prefix that will be used to lookup for injected secrets at runtime. This can be overriden to using
+FLYTE_SECRETS_ENV_PREFIX variable
+"""
+
+SECRETS_DEFAULT_DIR = _common_config.FlyteStringConfigurationEntry("secrets", "default_dir", default="/etc/secrets")
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flytekit/configuration/secrets.py:10]: for /etc/secrets you mean?

File: flytekit/models/security.py (hunk 5 above)
Comment (kumare3) [reply flytekit/models/security.py:34]: ANY what?

**Review by wild-endeavor — DISMISSED** [review]

**Review by wild-endeavor — APPROVED** [review]

File: flytekit/configuration/secrets.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+# Secrets management
+SECRETS_ENV_PREFIX = _common_config.FlyteStringConfigurationEntry("secrets", "env_prefix", default="_FSEC_")
+"""
+This is the prefix that will be used to lookup for injected secrets at runtime. This can be overriden to using
+FLYTE_SECRETS_ENV_PREFIX variable
+"""
+
+SECRETS_DEFAULT_DIR = _common_config.FlyteStringConfigurationEntry(
+    "secrets", "default_dir", default=os.path.join("etc", "secrets")
```
Comment (EngHabu) [inline flytekit/configuration/secrets.py:13]:
```suggestion
    "secrets", "default_dir", default=os.path.join(os.path.sep, "etc", "secrets")
```

**Review by EngHabu — APPROVED** [review]
